add Impedance task

mmurooka opened the pull request:
Add the task to move the end-effector with impedance dynamics. Unlike AdmittanceTask  or DampingTask, a robot can adapt to the external force while following the target position. For example, ImpedanceTask is necessary when pushing an object with a constant force (both force and position targets need to be considered).

This impedance task corresponds to what @stephane-caron  calls "impedance control" below: https://gite.lirmm.fr/multi-contact/mc_rtc/-/merge_requests/133#note_9535 (this link is internal only)

The movie of the sample controller (MCImpedanceController) added in this PR: https://www.dropbox.com/s/cjkdb9u5ihyotnz/impedance-controller-20201205.mp4?dl=0

mmurooka commented:
I add two notes about the force control of mc_rtc.

#### 1. Relationship between ImpedanceTask and AdmittanceTask

As shown in the attached PDF, the current AdmittanceTask in mc_rtc (and the force control presented in [this paper](https://hal.archives-ouvertes.fr/hal-01495662)) are a special case of impedance control. I wrote a small document about this, see [AdmittanceTask_mc_rtc_20201206.pdf](https://github.com/jrl-umi3218/mc_rtc/files/5648488/AdmittanceTask_mc_rtc_20201206.pdf).

#### 2. Class hierarchy of force control tasks

Because AdmittanceTask, DampingTask, and ComplianceTask are all special cases of ImpedanceTask, I suggest the following better class hierarchy. (Furthermore, AdmittanceTask and DampingTask do almost the same thing, so I don't think they need to be separated.)
But this is outside the scope of this Pull Request as it will affect the existing codes.

![class-hierarchy](https://user-images.githubusercontent.com/6636600/101275655-91a02a80-37ea-11eb-97ea-e04d105a64a4.png)

These notes are almost same as suggested in the internal MergeRequest linked in the first comment.

arntanguy commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+  sva::PTransformd T_0_c(compPoseW_.rotation());
+  // Represent the compliance velocity and acceleration in the compliance frame and scale by dt
+  sva::MotionVecd compDeltaC = T_0_c * (dt * compVelW_ + 0.5 * dt * dt * compAccelW_);
+  // Convert the angular velocity to the rotation matrix through AngleAxis representation
+  Eigen::AngleAxisd aaCompDeltaC(Eigen::Quaterniond::Identity());
+  if(compDeltaC.angular().norm() > 1e-6)
+  {
+    aaCompDeltaC = Eigen::AngleAxisd(compDeltaC.angular().norm(), compDeltaC.angular().normalized());
+  }
```
I think it would be nice to have a more direct way to do this kind of integration in `SpaceVecAlg`. I'm not sure how to best implement it though (ping @mehdi-benallegue)

arntanguy commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+
+  gui.addElement({"Tasks", name_},
+                 // pose
+                 mc_rtc::gui::Transform("desiredPose", [this]() { return this->desiredPose(); },
```
```suggestion
                 mc_rtc::gui::Transform("desiredPose", [this]() -> const sva::PTransformd & { return this->desiredPose(); },
```

arntanguy commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+                 // pose
+                 mc_rtc::gui::Transform("desiredPose", [this]() { return this->desiredPose(); },
+                                        [this](const sva::PTransformd & pos) { this->desiredPose(pos); }),
+                 mc_rtc::gui::Transform("compliancePose", [this]() { return this->compliancePose(); }),
```
```suggestion
                 mc_rtc::gui::Transform("compliancePose", [this]() -> const sva::PTransformd & { return this->compliancePose(); }),
```

mmurooka replied:
Because of the change after your review (sorry for that), this is no longer reference: https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/40265e6d8c428f77c3924a09ab3955f37c51e8f7#diff-db3809a3a52f623c53b2c5836c7546b76608fe22dd5e5e097bfc2c1e306e551dR215 .

arntanguy commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+: SurfaceTransformTask(surfaceName, robots, robotIndex, stiffness, weight)
+{
+  const auto & robot = robots.robot(robotIndex);
+  if(!robot.surfaceHasIndirectForceSensor(surfaceName))
+  {
+    mc_rtc::log::error_and_throw<std::runtime_error>(
+        "[mc_tasks::ImpedanceTask] Surface {} does not have a force sensor attached", surfaceName);
+  }
+
+  type_ = "impedance";
+  name_ = "impedance_" + robots.robot(rIndex).name() + "_" + surfaceName;
```
It's a bit nicer to use the more complete task name in the error here (it can sometimes be a bit tricky to track back where exception arose, so the more information the better):
```suggestion
  type_ = "impedance";
  name_ = "impedance_" + robots.robot(rIndex).name() + "_" + surfaceName;
  
  if(!robot.surfaceHasIndirectForceSensor(surfaceName))
  {
    mc_rtc::log::error_and_throw<std::runtime_error>(
        "[{}] Surface {} does not have a force sensor attached", name_, surfaceName);
  }

```

arntanguy reviewed (changes requested):
Thanks @mmurooka !

Overall this looks good, just a few minor comments and suggestions. You can commit most of them directly through the github interface ("Add suggestion to batch", and then "Commit Suggestion").

I just have a minor concern about the inconsistent units of impM/impD/impK (kg in the constructor, newtons in the setters).

Then let's try this on the robot ;)

mmurooka commented on src/mc_control/samples/Impedance/mc_impedance_controller.cpp:
```diff
@@ -0,0 +1,60 @@
+  rootLinkTask =
+      std::make_shared<mc_tasks::EndEffectorTask>("base_link", robots(), robots().robotIndex(), 100.0, 10000.0);
+  rootLinkTask->set_ef_pose(sva::PTransformd(Eigen::Vector3d(0, 0, 1)));
+  solver().addTask(rootLinkTask);
```
> Why a task on the root link?

Not so much reason. Just to make a example with most simple settings, the arm of JVRC1 is considered as a manipulator fixed to the ground.

Instead of it, adding foot/floor contact, dynamicsConstraint, CoM task is also good for humanoid example.

mmurooka commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+  // Set impedance parameters
+  impedance(sva::ForceVecd(Eigen::Vector3d::Constant(10.0 * impM), Eigen::Vector3d::Constant(impM)),
+            sva::ForceVecd(Eigen::Vector3d::Constant(10.0 * impD), Eigen::Vector3d::Constant(impD)),
+            sva::ForceVecd(Eigen::Vector3d::Constant(10.0 * impK), Eigen::Vector3d::Constant(impK)));
```
`impM/impD/impK` are `sva::ForceVecd` instance and the unit is different for the force part and moment part. For example, the unit of `impM`'s force part is `kg`, and the unit of `impM`'s moment part is `kg m^2` (same with moment of inertia). Therefore, the impedance parameters of the force and moment components cannot be compared.

Ten times comes from the experience of being too sensitive to moments when the same impedance parameters are applied to both forces and moments. Multiplying all of MDK by 10 makes it less insensitive to external moment.

mmurooka replied:
Updated in https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/0df36be714b2471c7abb5bc8156a754c683dfebb .
Enable to set the parameter scale of the orientation via constructor argument and make the safe value default (as we talked directly).

mmurooka reviewed (commented):
Thank you for your review!

mmurooka commented:
Thanks to the advice from @mitsuharu-morisawa, I updated the impedance equation implementation a little bit after the first review: https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/40265e6d8c428f77c3924a09ab3955f37c51e8f7 .

There is no change in behavior before and after the change if deisred pose, velocity, and acceleration are consistent. If desired pose, velocity, and acceleration are NOT consistent (e.g., desired pose is updated with ZERO desired velocity and acceleration), the new implementation, which deals only with the amount of modification (`deltaComp****W_`) in the impedance equation, more robustly follows the desired pose.

mmurooka commented on include/mc_tasks/ImpedanceTask.h:
```diff
@@ -0,0 +1,285 @@
+   *  See the Constructor description for the definition of compliance pose.
+   *
+   */
+  const sva::PTransformd & compliancePose() const
```
The implementation is changed after review and no longer `noexcept`.

arntanguy commented:
I added the JSON schema, we can now add and test the task directly from the GUI.

Note that we will need to do a few code changes later on, as the YAML properties inherited from `SurfaceTransformTask` are not all valid as-is (they call `target()` instead of your `desiredPose(...)` for instance).

mmurooka commented on include/mc_tasks/ImpedanceTask.h:
```diff
@@ -0,0 +1,285 @@
+  // surface frame.
+  sva::PTransformd compPoseW_ = sva::PTransformd::Identity();
+  sva::MotionVecd compVelW_ = sva::MotionVecd::Zero();
+  sva::MotionVecd compAccelW_ = sva::MotionVecd::Zero();
```
updated by https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/ac2574bc0b5bfa836f901946ae4d016d67bb4a72 .

mmurooka commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+                // Compute in the surface frame because the impedance parameters and wrench gain are represented in the
+                // surface frame
+                impM_.vector().cwiseInverse().cwiseProduct(
+                    // T_0_s transforms the MotionVecd value from the world from to the surface frame
```
updated by https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/ac2574bc0b5bfa836f901946ae4d016d67bb4a72 .

mmurooka commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,238 @@
+  // 2.1 Integrate velocity to pose
+  sva::PTransformd T_0_c(compPoseW_.rotation());
+  // Represent the compliance velocity and acceleration in the compliance frame and scale by dt
+  sva::MotionVecd compDeltaC = T_0_c * (dt * compVelW_ + 0.5 * dt * dt * compAccelW_);
```
updated by https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/ac2574bc0b5bfa836f901946ae4d016d67bb4a72 and https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/3d59973e7a3e4392a16d27afc84e6d2637984004 .

mmurooka commented:
> I added the JSON schema, we can now add and test the task directly from the GUI. Note that we will need to do a few code changes later on

Thanks! I could add a new ImpedanceTask from Rviz panel. It seems that we don't need any  code changes ;)
https://www.dropbox.com/s/wh6z3ijpmcggp9n/impedance-task-from-rviz-20201210.mp4?dl=0

mmurooka commented:
- [ ] clamp gains (especially non-zero for impedance M to avoid zero devision).

- [ ] clamp the compliance delta pose.

- [ ] set the initial target from yaml.

- [ ] enable to set consistent impedance MDK.

mmurooka commented:
CI seems to start to fail with some issue unrelated to PR..

https://github.com/jrl-umi3218/mc_rtc/pull/100/checks?check_run_id=1554274341#step:5:1556
```bash
/package/mc_rtc/src/mc_rbdyn/Robot.cpp:1236:5: error: 'Quaterniond' is not a member of 'sva'
     sva::Quaterniond rotation{pt.rotation().transpose()};
     ^
/package/mc_rtc/src/mc_rbdyn/Robot.cpp:1236:5: note: suggested alternative:
In file included from /usr/include/eigen3/Eigen/Geometry:42:0,
                 from /usr/include/SpaceVecAlg/SpaceVecAlg:13,
                 from /usr/include/RBDyn/Joint.h:16,
                 from /package/mc_rtc/include/mc_rbdyn/Base.h:9,
                 from /package/mc_rtc/include/mc_rbdyn/Robot.h:7,
                 from /package/mc_rtc/src/mc_rbdyn/Robot.cpp:5:
/usr/include/eigen3/Eigen/src/Geometry/Quaternion.h:303:28: note:   'Eigen::Quaterniond'
 typedef Quaternion<double> Quaterniond;
                            ^
```

gergondet commented:
> CI seems to start to fail with some issue unrelated to PR..
> 
> [#100 (checks)](https://github.com/jrl-umi3218/mc_rtc/pull/100/checks?check_run_id=1554274341#step:5:1556)

@mmurooka Yes. This is related to jrl-umi3218/SpaceVecAlg#37 and fixed by a13a66f4934010cc5086ccf983f7918156a22e52 so if you rebase on master it should be fine

arntanguy commented:
Pierre was faster than me there, I've rebased it onto master and will now look into the last steps to finalize this PR :)

arntanguy commented:
This is almost finalized:
- 5466479 adds a fixed-based variant of the `JVRC1` robot module named `JVRC1Fixed`
- 97ea60e improves the sample controller to use the fixed-based robot, and make the trajectory start smoothly at the current position
- 5c375f4 Fixes a numerical issue with the BodySensor observer that only occurs with fixed-based robots: when repeatedly updating the pose of the floating base, the computation of relative transformation between the IMU sensor and the floating base becomes de-orthogonalized and de-normalized. This could be avoided by computing this relative pose without passing though world-frame, I'll introduce this later. For now this reorthogonalizes the transformation. (this doesn't directly affect this PR but affects the mc_force_sensor_calibration controller)
- 40aee1d correctly loads the task's target from YAML. I went the simplest way of setting the compliance pose the the same as the `TrajectoryTaskGeneric`'s one.
- 5676705 adds a tutorial page about all the available sample controllers, with relevant links to the corresponding API/YAML/other tutorials/online demos.

TODO:
- [ ] Make the mc_force_sensor_calibration controller public

mmurooka commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,316 @@
+
+  if(config.has("target"))
+  {
+    desiredPose(config("target"));
```
Maybe we should remove this part to remove the conflict with 
https://github.com/jrl-umi3218/mc_rtc/pull/100/files#diff-1ea7af2780604d3059658bfd0c8f69e98c1a603be1810ec9c1828df358f272a1R205

arntanguy replied:
Nice catch, done in 9558f32

mmurooka reviewed (commented):
Thank you for finalizing!

arntanguy commented:
Hop all done, and extensively tested on HRP5 (with both the bobbin, and the stabilizer in #95).

I think this can be merged when CI passes.
@gergondet Any further comments on this PR?

gergondet commented on src/mc_robots/jvrc1.h:
```diff
@@ -14,7 +14,35 @@ namespace mc_robots
+      return nullptr;
+    }
+  }
+}
```
This breaks `MC_RTC_BUILD_STATIC`, you can have (to be passed through clang-format):

```cpp
#ifndef MC_RTC_BUILD_STATIC
extern "C"
{
  ROBOT_MODULE_API void MC_RTC_ROBOT_MODULE(std::vector<std::string> & names)
  {
    names = {"JVRC1", "JVRC1Fixed"};
  }
  ROBOT_MODULE_API void destroy(mc_rbdyn::RobotModule * ptr)
  {
    delete ptr;
  }
  ROBOT_MODULE_API mc_rbdyn::RobotModule * create(const std::string & name)
  {
    if(name == "JVRC1")
    {
      return new mc_robots::JVRC1RobotModule(false);
    }
    else if(name == "JVRC1Fixed")
    {
      return new mc_robots::JVRC1RobotModule(true);
    }
    else
    {
      mc_rtc::log::error("JVRC1 module Cannot create an object of type {}", name);
      return nullptr;
    }
  }
}
#else
namespace                                                                          
{                                                                                  
    static auto registered = []() {                                                    
      using fn_t = std::function<JVRC1RobotModule *()>;                                            
      mc_rbdyn::RobotLoader::register_object("JVRC1", fn_t([]() { return new JVRC1RobotModule(false); })); 
      mc_rbdyn::RobotLoader::register_object("JVRC1Fixed", fn_t([]() { return new JVRC1RobotModule(true); })); 
      return true;                                                                     
    }();                                                                               
}
#endif

gergondet commented on include/mc_tasks/ImpedanceTask.h:
```diff
@@ -0,0 +1,359 @@
+  sva::ForceVecd impM_; // must be set in the Constructor
+  sva::ForceVecd impD_; // must be set in the Constructor
+  sva::ForceVecd impK_; // must be set in the Constructor
+  sva::MotionVecd wrenchGain_ = sva::MotionVecd::Zero();
```
[`sva::ImpedanceVecd`](https://github.com/jrl-umi3218/SpaceVecAlg/blob/master/src/SpaceVecAlg/ImpedanceVec.h) is a better match for this kind of parameters, it's similar to `sva::ForceVecd` but has `linear()` and `angular()` accessors which are more explicit. Although they are not super well supported by mc_rtc so I will make a PR for this later today.

gergondet replied:
> Although they are not super well supported by mc_rtc so I will make a PR for this later today.

This is done in #112

gergondet commented on include/mc_tasks/ImpedanceTask.h:
```diff
@@ -0,0 +1,359 @@
+  }
+
+  /*! \brief Get the desired pose of the surface in the world frame. */
+  const sva::PTransformd & desiredPose() const noexcept
```
We use `target` rather than `desired` everywhere else so for coherency it would be better to use `target` here as well (this comment applies to every use of `desired` in this class and the associated schema)

mmurooka replied:
Originally, I intentionally used a different term than target. The desiredPose specified by the user is modified by the impedance equation, and it is passed to the target of the SurfaceTransformTask: https://github.com/mmurooka/mc_rtc/blob/519a88a24b65b0bd7472dfca4b0833aaba1d74c3/src/mc_tasks/ImpedanceTask.cpp#L141

I thought that overriding the target method in ImpedanceTask would cause the following two return values to be different, which would be confusing:
```
std::shared_ptr<ImpedanceTask> task;
task->target();
std::dynamic_pointer_cast<SurfaceTransformTask>(task)->target();
```

However, it doesn't matter in normal usage, and target may be easier for the user to understand, so either one is fine.

gergondet commented on include/mc_tasks/ImpedanceTask.h:
```diff
@@ -0,0 +1,359 @@
+        mc_filter::utils::clampAndWarn(gain.linear(), wrenchGainLinLower_, wrenchGainLinUpper_, "wrenchGainLinear");
+    wrenchGain_.angular() =
+        mc_filter::utils::clampAndWarn(gain.angular(), wrenchGainAngLower_, wrenchGainAngUpper_, "wrenchGainAngular");
+  }
```
How about adding an `mc_tasks::force::ImpedanceGains` class to lighten the `mc_tasks::force::AdmittanceTask` API?

Something like (simplified for the review):
```cpp
struct ImpedanceGains
{
  /* relevant accessors */
protected:
  // Either sva::ImpedanceVecd or an internal wrapper that ensures the gains are (strictly) positive
  sva::ImpedanceVecd M_;
  sva::ImpedanceVecd D_;
  sva::ImpedanceVecd K_;
  sva::ImpedanceVecd wrench_;
};
```

```cpp
mc_tasks::force::AdmittanceTask task;
// For example, instead of:
task.impedanceK(...);
// We get
task.gains().K(...);
// And with a nice wrapper for ensuring the provided gains are positive we could have:
task.gains().K().linear(...);
// Without having to add as many functions in ImpedanceTask itself
```

Down the line, if we factorize the other force tasks as suggested we can specialized the `ImpedanceGains` accordingly and expose a specific `gains()` method for each force task rather than painstakingly choosing and picking specific methods

gergondet commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,311 @@
+  if(lowPass_.dt() != solver.dt())
+  {
+    lowPass_.dt(solver.dt());
+  }
```
You can override `addToSolver` to do this only once

mmurooka replied:
Thank you for telling nicer approach. Fixed in https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/42ebb40896e82c93aca68902660b0c97032cb2ce.

gergondet commented on src/mc_tasks/ImpedanceTask.cpp:
```diff
@@ -0,0 +1,311 @@
+                 mc_rtc::gui::Transform("pose", [this]() { return this->surfacePose(); }),
+                 // impedance parameters
+                 mc_rtc::gui::ArrayInput("impedanceM", {"cx", "cy", "cz", "fx", "fy", "fz"},
+                                         [this]() { return this->impedanceM().vector(); },
```
FYI, you can return a `const sva::ForceVecd &` here as the GUI serializer knows how to transform it into an array

gergondet replied:
And with #112 it's actually a little better to do so as this incurs no additional copy to get the vector

mmurooka replied:
> FYI, you can return a const sva::ForceVecd & here as the GUI serializer knows how to transform it into an array

Although `ForceVecd::vector()` does not return const reference, can we return const reference in GUI?
https://github.com/jrl-umi3218/SpaceVecAlg/blob/1668efd3f2647c6a850d268b73ce21930ebbeb38/src/SpaceVecAlg/ForceVec.h#L88

gergondet reviewed (changes requested):
Looks good but I would like to clean up the API a bit before we merge this

mmurooka commented on include/mc_tasks/ImpedanceTask.h:
```diff
@@ -0,0 +1,359 @@
+  const Eigen::Vector3d wrenchGainLinUpper_ = Eigen::Vector3d::Constant(1e6);
+  const Eigen::Vector3d wrenchGainLinLower_ = Eigen::Vector3d::Constant(0);
+  const Eigen::Vector3d wrenchGainAngUpper_ = Eigen::Vector3d::Constant(1e6);
+  const Eigen::Vector3d wrenchGainAngLower_ = Eigen::Vector3d::Constant(0);
```
Yes, I think the upper bounds may be omitted for all variables.
As for lower bounds, it should be strictly positive for `M`, and zero or positive for other variables, but I think separate limits are unnecessary.

mmurooka commented:
@gergondet , Thank you for review!
I generally agree with the comments. However, I'm not sure if I can reflect your comments exactly as you intended, so I would appreciate it if you could change some (especially about https://github.com/jrl-umi3218/mc_rtc/pull/100#discussion_r551092128) or all directly when you have time.

gergondet commented:
@mmurooka thanks for your answer, I've pushed what I had in mind for `ImpedanceGains`, you can see it in [`3037f48` (#100)](https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/3037f48d20809c8f598310ed0382603cf89eed2c)

I went ahead with a few other changes discussed in the PR. Could you test the changes in a simulation? I don't think I have the corresponding JVRC1 environment for testing the sample.

mmurooka commented:
Thank you for update. I tried Impedance controller in choreonoid but, it seems something has broken...
video: https://www.dropbox.com/s/n3fy3ykakjb4na6/impedance_20210107.mp4?dl=0

During simulation, the following warnings continue to be output:
```
[warning] angular deltaCompAccel limited from 2788.7463790536262 to 1000.0
[warning] linear deltaCompPose limited from 1.0001272017767997 to 1.0
[warning] linear deltaCompAccel limited from 2457.9901027281644 to 1000.0
```

> I don't think I have the corresponding JVRC1 environment for testing the sample.

I thought that it is just doing
```
cd ~/openrtp/share/hrpsys/samples/JVRC1 # or corresponding directory
choreonoid sim_mc_fixed.cnoid --start-simulation
```
with the following mc_rtc.yaml
```
MainRobot: JVRC1Fixed
Enabled: Impedance
```
Is there anything missing to reproduce this in your side?

gergondet commented:
@mmurooka Indeed I have everything to test it here

I have now fixed the sample behavior for both the fixed base and floating base variant of JVRC1

For reference, I had mistakenly inverted the angular/linear gains [`4f82a63` (#100)](https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/4f82a6306dc93702ab53157d4b58490d2770735a)

And more importantly, `spring()` was not returning the correct gain at all [`328c6cd` (#100)](https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/328c6cd3f9ba6be1473186b1768819b476142dd3)

It should be ok now

mmurooka commented:
@gergondet Thank you. I now checked a simulation works correctly for fixed and non-fixed JVRC1. I read through the changes in the source code and it's OK.

I changed the damping parameter of sample controller because the impedance vibrates and diverges when adding large force interactively by clicking and dragging in Choreonoid (also for the real robot, the critical damping value does not work well for ImpedanceTask).

Only one concern is that the default impedance gain can be dangerous in some cases.
We have encountered divergence of ImpedanceTask on real robots many times (e.g., https://gite.lirmm.fr/jrp-michelin/mc_hwm_planner/-/issues/7) and I'd like to keep the source code safe side.

If we change the wrench gain to 1 (it is zero for default) with keeping MDK with initial value (M=1e-6 and D,K=0), it diverges. It could easily happen by user's operation for example when adding task interactively like https://www.dropbox.com/s/wh6z3ijpmcggp9n/impedance-task-from-rviz-20201210.mp4?dl=0 (this link is copied from https://github.com/jrl-umi3218/mc_rtc/pull/100#issuecomment-742527233)

gergondet commented:
Thanks @mmurooka , that's a great point. I've changed the default gains in [`71b04b9` (#100)](https://github.com/jrl-umi3218/mc_rtc/pull/100/commits/71b04b9ef939aa523cc1d167dbf410b6957a7358) so that they match the ones you were setting before.

I'm also thinking of changing the GUI of the gains so that they can be set in a bulk rather than having to change each parameters

mmurooka commented:
I added small two commits (Just in case it is not erased by force push).

> I've changed the default gains in 71b04b9 (#100) so that they match the ones you were setting before.

Thank you, I confirmed it worked as intended.

> I'm also thinking of changing the GUI of the gains so that they can be set in a bulk rather than having to change each parameters

It will be useful (although we also need to keep the element-wise interface).

gergondet commented:
> It will be useful (although we also need to keep the element-wise interface).

This looks like this in the latest change (with a patch to mc_rtc_ros I just published):

![image](https://user-images.githubusercontent.com/473543/103971982-edbdec00-51a6-11eb-8636-091c8ec6c2b4.png)

You can still edit a single gain but they are also all sent (and changed) at the same time

gergondet commented:
> You can still edit a single gain but they are also all sent (and changed) at the same time

Quick note that just came to me: forms do not (yet) support having their default data fed from a callbacks so I will actually revert that change since the GUI would not reflect changes made programmatically to the gains

mmurooka commented:
OK, then I think we can leave it as it is. (And also I thought the order of elements (i.e., angular followed by linear) is not clear for beginners in the form style).
